### PR TITLE
Fix padded_gather to allow tensor attributes in TrainResult

### LINKED ISSUE #3167: Can't set result object predictions to tensors
## 🐛 Bug

[The docs ](https://pytorch-lightning.readthedocs.io/en/0.9.0/results.html) state that "If you’d like to do something special with the outputs other than logging, implement __epoch_end." and gives the following example:

```python
def training_step(self, batch, batch_idx):
    result = pl.TrainResult(loss)
    result.some_prediction = some_prediction
    return result

def training_epoch_end(self, training_step_output_result):
    all_train_predictions = training_step_output_result.some_prediction

    training_step_output_result.some_new_prediction = some_new_prediction
    return training_step_output_result
```

When the custom value is a Tensor, this usage fails in Result.padded_gather() method where all entries in the result are traversed, and their `tbptt_pad_token` meta property is checked, since the custom property doesn't have meta:

```
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\states.py", line 48, in wrapped_fn
    result = fn(self, *args, **kwargs)
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 1073, in fit
    results = self.accelerator_backend.train(model)
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\accelerators\gpu_backend.py", line 51, in train
    results = self.trainer.run_pretrain_routine(model)
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\trainer.py", line 1239, in run_pretrain_routine
    self.train()
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\training_loop.py", line 394, in train
    self.run_training_epoch()
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\training_loop.py", line 550, in run_training_epoch
    self.run_training_epoch_end(epoch_output, checkpoint_accumulator, early_stopping_accumulator, num_optimizers)
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\training_loop.py", line 657, in run_training_epoch_end
    epoch_output = self.__gather_result_across_time_and_optimizers(epoch_output)
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\trainer\training_loop.py", line 733, in __gather_result_across_time_and_optimizers
    gathered_opt_output = time_gathered_outputs[0].__class__.padded_gather(time_gathered_outputs)
  File "C:\Users\yigit\Anaconda3\lib\site-packages\pytorch_lightning\core\step_result.py", line 338, in padded_gather
    padding_key = meta[name]['tbptt_pad_token']
KeyError: 'probs'
```

### To Reproduce

Try train/val loop example with custom attribute in [the Result objects documentation](https://pytorch-lightning.readthedocs.io/en/0.9.0/results.html).

### Expected behavior

I don't know what padding in Result does, but I guess it must be applied to logged results only.

### Environment

* Packages:
        - numpy:             1.18.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.6.0
        - pytorch-lightning: 0.9.0
        - tensorboard:       2.2.0
        - tqdm:              4.47.0
* System:
        - OS:                Windows
        - architecture:
                - 64bit
                - WindowsPE
        - processor:         Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
        - python:            3.7.1
        - version:           10.0.19041

### LINKED ISSUE #3356: Inconsistent behavior between `validation_epoch_end` and `training_epoch_end`
## Inconsistent behavior between `validation_epoch_end` and `training_epoch_end`

Assignation of a variable to results in train_step (`result.y =y`) does not work the same as in validation. Take a look at the code in the section below to make it more clear.

For context, I'm trying to calculate AUC after both training and validation epochs. Calculating on batch might not work if the particular batch gets labels for only one class, so it would be undefined for the particular step.

### To Reproduce
Include this in a `pl` module:

```python
    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.criterion(y_hat, y)
        result = pl.TrainResult(loss)
        result.log('train_loss', loss)
        
        
        y_pred = F.softmax(y_hat, dim=0)[:, 1]

        result.y = y
        result.y_pred = y_pred

        return result
  
    def training_epoch_end(self, outputs):
        y = outputs.y
        y_pred = outputs.y_pred

        roc_auc_fn = pl.metrics.classification.AUROC()
        roc_auc = roc_auc_fn(y_pred, y)

        result = pl.TrainResult()
        result.log('train_roc_auc', roc_auc)

        return result
```

Then the stacktrace shows:

```
---------------------------------------------------------------------------

KeyError                                  Traceback (most recent call last)

<ipython-input-48-01044cd5ab91> in <module>()
      7 # train
      8 trainer = pl.Trainer(gpus=1, max_epochs=50)
----> 9 trainer.fit(model, dm)

8 frames

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/states.py in wrapped_fn(self, *args, **kwargs)
     46             if entering is not None:
     47                 self._state = entering
---> 48             result = fn(self, *args, **kwargs)
     49 
     50             # The INTERRUPTED state can be set inside the run function. To indicate that run was interrupted

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/trainer.py in fit(self, model, train_dataloader, val_dataloaders, datamodule)
   1071         """
   1072         # --------------------------
-> 1073         # Setup??
   1074         # --------------------------
   1075         ref_model = model

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/accelerators/gpu_backend.py in train(self, model)
     49 
     50         self.trainer.model = model
---> 51 
     52     def train(self):
     53         model = self.trainer.model

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/trainer.py in run_pretrain_routine(self, model)
   1237             trainer = Trainer()
   1238             model = LightningModule()
-> 1239 
   1240             trainer.fit(model)
   1241             trainer.test(test_dataloaders=test, ckpt_path='path/to/checkpoint.ckpt')

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/training_loop.py in train(self)
    392 
    393                 # early stopping
--> 394                 met_min_epochs = epoch >= self.min_epochs - 1
    395                 met_min_steps = self.global_step >= self.min_steps if self.min_steps else True
    396 

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/training_loop.py in run_training_epoch(self)
    548         # TODO bake this logic into the checkpoint callback
    549         should_activate = not is_overridden('validation_step', self.get_model()) and not should_check_val
--> 550         if should_activate:
    551             checkpoint_callbacks = [c for c in self.callbacks if isinstance(c, ModelCheckpoint)]
    552             [c.on_validation_end(self, self.get_model()) for c in checkpoint_callbacks]

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/training_loop.py in run_training_epoch_end(self, epoch_output, checkpoint_accumulator, early_stopping_accumulator, num_optimizers)
    655         epoch_log_metrics = {}
    656         epoch_progress_bar_metrics = {}
--> 657         for opt_outputs in epoch_output:
    658             # reduce across time first
    659             time_reduced_outputs = []

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/trainer/training_loop.py in __gather_result_across_time_and_optimizers(self, epoch_output)
    731         is_last_batch_for_infinite_dataset = (is_last_batch and self.val_check_batch == float('inf'))
    732         should_check_val = can_check_val and (should_check_val or is_last_batch_for_infinite_dataset)
--> 733 
    734         return should_check_val
    735 

/usr/local/lib/python3.6/dist-packages/pytorch_lightning/core/step_result.py in padded_gather(cls, outputs)
    336 
    337                 if is_reserved:
--> 338                     padding_key = default_padding_idx
    339                 else:
    340                     padding_key = meta[name]['tbptt_pad_token']

KeyError: 'y'
```

#### Code sample
Full module to give full context.

```python
class LitModel(pl.LightningModule):

    def set_parameter_requires_grad(self):
        if self.feature_extract:
            for param in self.model.parameters():
                param.requires_grad = False

    def __init__(self, num_classes, use_pretrained=True, feature_extract = True):
        super().__init__()
        self.feature_extract = feature_extract
        self.criterion = nn.CrossEntropyLoss()

        self.model = models.resnet18(pretrained=use_pretrained)
        self.set_parameter_requires_grad()

        num_ftrs = self.model.fc.in_features
        self.model.fc = nn.Linear(num_ftrs, num_classes)

        self.input_size = 224

    def forward(self, x):
        x = self.model(x)
        return x

    def configure_optimizers(self):
        print("Params to learn:")
        if self.feature_extract:
            params_to_update = []
            for name, param in self.model.named_parameters():
                if param.requires_grad == True:
                    params_to_update.append(param)
                    print("\t",name)
        else:
            params_to_update = self.model.parameters()
            for name,param in model_ft.named_parameters():
                if param.requires_grad == True:
                    print("\t",name)

        optimizer = optim.Adam(params_to_update, lr=0.0001, betas=(0.9, 0.999), eps=1e-08, weight_decay=0)
        return optimizer

    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.criterion(y_hat, y)
        result = pl.TrainResult(loss)
        result.log('train_loss', loss)
        
        
        y_pred = F.softmax(y_hat, dim=0)[:, 1]

        result.y = y
        result.y_pred = y_pred

        return result

    def training_epoch_end(self, outputs):
        y = outputs.y
        y_pred = outputs.y_pred

        roc_auc_fn = pl.metrics.classification.AUROC()
        roc_auc = roc_auc_fn(y_pred, y)

        result = pl.TrainResult()
        result.log('train_roc_auc', roc_auc)

        return result

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = self.criterion(y_hat, y)
        result = pl.EvalResult(checkpoint_on=loss)
        result.log('val_loss', loss)

        y_pred = F.softmax(y_hat, dim=0)[:, 1]

        result.y = y
        result.y_pred = y_pred

        return result

    def validation_epoch_end(self, outputs):
        y = outputs.y
        y_pred = outputs.y_pred

        roc_auc_fn = pl.metrics.classification.AUROC()
        roc_auc = roc_auc_fn(y_pred, y)

        result = pl.EvalResult(checkpoint_on=roc_auc)
        result.log('val_roc_auc', roc_auc)

        return result
```

### Expected behavior

I would expect the behavior in both to be quite similar, in particular, if both `*_epoch_end` methods have the same signatures. The current docs do not mention how both methods differ so I interpreted you could do the same.

### Environment

pytorch lightning on colab, installed wit
pytroch lightning version 0.9.0

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:

 - PyTorch Version (e.g., 1.0):1.6.0+cu101
 - OS (e.g., Linux): colab

### PR DESCRIPTION
**Pull request by teddykoker**
## What does this PR do?

Setting tensor attributes (see example below) to TrainResults caused a runtime error in `padded_gather` as these attributes are not in the `meta` dict. My fix has these attributes collated separately, or not at all if they are not tensors. 

Fixes #3356 
Fixes #3167

# Before submitting

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by teddykoker** [comment]
The issue here is with assigning attributes to `TrainResults`:
```python
def training_step(self, batch, batch_idx):
    x, y = batch
    y_hat = self(x)
    loss = F.cross_entropy(y_hat, y)
    result = TrainResult(minimize=loss)
    result.log_dict({"train_loss": loss})

    result.y = y
    result.y_hat = y_hat
    return result

def training_epoch_end(self, out):
    epoch_loss = F.cross_entropy(out.y_hat, out.y)
    result = TrainResult()
    result.log_dict({"train_epoch_loss": epoch_loss})
    return result
```
This should work (as shown in the docs)

**Comment by teddykoker** [comment]
These are the important changes 
https://github.com/PyTorchLightning/pytorch-lightning/pull/3530/files#diff-9471ab95794bb8e572901942af92359eL319-L349

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -306,111 +306,116 @@ def __copy__(self):
...
                 else:
-                    padding_key = meta[name]['tbptt_pad_token']
+                    padding_key = meta[name]["tbptt_pad_token"]
                 padded = torch.nn.utils.rnn.pad_sequence(value, batch_first=True, padding_value=padding_key)
                 result[name] = padded
 
                 # also update the result
                 if meta and not is_reserved:
-                    meta[name]['value'] = padded
+                    meta[name]["value"] = padded
```
Comment (teddykoker) [inline pytorch_lightning/core/step_result.py:345]: These are the changes

File: pytorch_lightning/core/step_result.py [hunk 2]
```diff
@@ -306,111 +306,116 @@ def __copy__(self):
...
+
+        # collate tensors that are not in meta (e.g. user assigned attributes)
+        for name, value in result.items():
+            if name not in meta:
+                result[name] = collate_tensors(result[name])
+
```
Comment (teddykoker) [inline pytorch_lightning/core/step_result.py:351]: This is the biggest change, user assigned attributes must be collated separately as they are not in the `meta` dict.

**Comment by Borda** [comment]
@teddykoker could you please set your Balck to ignore ' and " with `-S, --skip-string-normalization`

**Comment by teddykoker** [comment]
Truncated back propagation through time works failing ..., still WIP

File: tests/core/test_results.py [hunk 3]
```diff
@@ -240,7 +230,90 @@ def test_result_gather_mixed_types():
...
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        weights_summary=None,
+    )
+    trainer.fit(LitModel(), dm)
+
+
+def test_result_valid_epoch_end(tmpdir):
+    class LitModel(LightningModule):
```
Comment (Borda) [inline tests/core/test_results.py:278]: How this differs from our `EvalModelTemplate`, pls write just the methods which differ...

File: pytorch_lightning/core/step_result.py [hunk 4]
```diff
@@ -325,15 +325,16 @@ def padded_gather(cls, outputs):
         default_padding_idx = 0
         for name, value in result.items():
             if isinstance(value, list) and len(value) > 0 and isinstance(value[0], torch.Tensor):
-                if name not in {'checkpoint_on', 'early_stop_on', 'minimize'}:
+                if name not in {'checkpoint_on', 'early_stop_on', 'minimize'} and name in meta:
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:328]:
```suggestion
                is_reserved = name in {'checkpoint_on', 'early_stop_on', 'minimize'}
                if not is_reserved and name in meta:
```

**Review by Borda — CHANGES_REQUESTED** [review]
pls simplify the tests...

**Comment by edenlightning** [comment]
@teddykoker should we kill it?

**Comment by teddykoker** [comment]
@edenlightning lemme double check this isn’t necessary for the new log api otherwise I’ll kill it

**Pull request closed without merging**